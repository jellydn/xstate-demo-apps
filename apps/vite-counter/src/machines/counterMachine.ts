import { inspect } from "@xstate/inspect";
import { assign, createMachine, interpret } from "xstate";

if (typeof window !== "undefined") {
  inspect({
    iframe: false,
  });
}

type Ctx = {
  count: number;
};

const increment = (context: Ctx) => context.count + 1;
const decrement = (context: Ctx) => context.count - 1;

export const counterMachine =
  /** @xstate-layout N4IgpgJg5mDOIC5QGMD2BXAdgFzAJwAIBbAQ2QAsBLTMAOjO0oDcwBiASQDkBhRUAB1SxKjVJj4gAHogC0ARgBMABloBWAJyqFcpXIDsADgBsAFgWqANCACesueoO0DiuWYDMJ-RtUBfH1bQsXEJSCmo6BmY2ABEAUV4kEEFhUXFE6QR5ZTVNbV1DU3MrW0yFEyMnF3dPPW8-fxBMVAg4CUCcfGIyKhp6ZEYWCWSRSjEJDJk3HVo5VTl5gzdnPSNa4rsHSu1qr00-AIwOkO7woaERsfTZNw8ZuYWl-VXLG2v7Ldc9N00jc3U9eo+IA */
  createMachine(
    {
      predictableActionArguments: true,
      context: { count: 0 },
      tsTypes: {} as import("./counterMachine.typegen").Typegen0,
      schema: {
        context: {} as Ctx,
        events: {} as
          | {
              type: "INC";
            }
          | { type: "DEC" },
      },
      id: "counter machine",
      initial: "active",
      states: {
        active: {
          on: {
            INC: {
              actions: "increase",
            },
            DEC: {
              actions: "decrement",
            },
          },
        },
      },
    },
    {
      actions: {
        increase: assign({ count: increment }),
        decrement: assign({ count: decrement }),
      },
    }
  );

export const counterService = interpret(counterMachine, { devTools: true })
  .onTransition((state) => console.log(state.context.count))
  .start();
