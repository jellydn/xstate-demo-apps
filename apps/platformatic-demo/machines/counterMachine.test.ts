import { describe, expect, vi } from "vitest";
import { counterMachine, counterService } from "./counterMachine";

describe("counterMachine", () => {
  it("should increment counter", () => {
    const result = counterMachine.transition(
      counterMachine.initialState,
      "INC",
    );
    expect(result.context.count).toBe(1);
    expect(result).toMatchSnapshot();
  });
  it("should decrease counter", () => {
    const result = counterMachine.transition(
      counterMachine.initialState,
      "DEC",
    );
    expect(result.context.count).toBe(-1);
    expect(result).toMatchSnapshot();
  });
});

describe("counterService", () => {
  it("should increment counter", () => {
    const fn = vi.fn();

    counterService.subscribe(fn);
    counterService.send("INC");
    counterService.send("INC");

    expect(counterService.state).toMatchSnapshot();
    expect(counterService.state.context.count).toBe(2);
    expect(fn).toHaveBeenCalledTimes(3);
  });

  it("should decrease counter", () => {
    const fn = vi.fn();
    counterService.subscribe(fn);
    counterService.send("DEC");

    expect(counterService.state).toMatchSnapshot();
    expect(counterService.state.context.count).toBe(1);
    expect(fn).toHaveBeenCalledTimes(2);
  });
});
