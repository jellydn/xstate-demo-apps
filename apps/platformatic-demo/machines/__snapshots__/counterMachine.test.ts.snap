// Vitest Snapshot v1

exports[`counterMachine > should decrease counter 1`] = `
{
  "actions": [],
  "changed": true,
  "context": {
    "count": -1,
  },
  "matches": [Function],
  "value": "active",
}
`;

exports[`counterMachine > should increment counter 1`] = `
{
  "actions": [],
  "changed": true,
  "context": {
    "count": 1,
  },
  "matches": [Function],
  "value": "active",
}
`;

exports[`counterService > should decrease counter 1`] = `
{
  "actions": [],
  "changed": true,
  "context": {
    "count": 1,
  },
  "matches": [Function],
  "value": "active",
}
`;

exports[`counterService > should increment counter 1`] = `
{
  "actions": [],
  "changed": true,
  "context": {
    "count": 2,
  },
  "matches": [Function],
  "value": "active",
}
`;
